### ISSUE TITLE
Now getting `reportGeneralTypeIssues` when combining `divmod` with `typing.Self`

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6760]
### Discussed in https://github.com/microsoft/pyright/discussions/6757

<div type='discussions-op-text'>

<sup>Originally posted by **ncanceill** December 16, 2023</sup>
### Example

PyRight 1.1.331 was fine with this code but 1.1.336 is not:
```python
from typing import Self

class Test:
    def __divmod__(self, other: object) -> tuple[Self, Self]:
        ...

    def test(self, other: object) -> Self:
        (result, _) = divmod(self, other)
        return result  # Type "Test" cannot be assigned to type "Self@Test" (reportGeneralTypeIssues)
```

### Problem

Why did the behavior change?
- The code is actually fine and this is a bug/regression
- The code must be written differently for some reason
- The code actually has wrong typing for some reason
- Other

### Notes

The changelog for [1.1.335](https://github.com/microsoft/pyright/releases/tag/1.1.335) says:
> Fixed bug that led to a false negative when assigning type `Class` to `Self@Class`

And the changelog for [1.1.336](https://github.com/microsoft/pyright/releases/tag/1.1.336) says:
> Fixed a bug that leads an incorrect type evaluation when binding a magic method (like `__add__`) to `self` when the magic method returns `Self`.

### More examples

PyRight 1.1.336 passes when using the magic method instead of the built-in function:
```python
class Test:
    def __divmod__(self, other: object) -> tuple[Self, Self]:
        ...
    
    def test_(self, other: object) -> Self:
        (result, _) = self.__divmod__(other)
        return result
```

And it also passes when replacing `__divmod__` with the `__add__` magic method:
```python
class Test:
    def __add__(self, other: object) -> Self:
        ...
    
    def test_(self, other: object) -> Self:
        result = self + other
        return result
```
</div>

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.342, which I just published. It will also be included in a future release of pylance.
